### PR TITLE
kernel: Divide Event into ReadableEvent and WritableEvent

### PR DESCRIPTION
**Pull request by DarkLordZach**
Our current event implementation is not very hardware accurate. Previously, we just had a singular Event class that did everything related to events, but on actual hardware there are two, ReadableEvent and WritableEvent.

From a high-level perspective, these two classes have a very similar relationship compared to that of ServerPort and ClientPort. Sysmodules/services will create an event pair and register the readable event with the kernel. When asked for an event (via and IPC command), sysmodules return the ReadableEvent. Modules can then use the WritableEvent to signal/clear the event.

This addresses part of the issues in #2.

### CODE REVIEW

File: src/core/hle/service/am/am.cpp [hunk 1]
```diff
@@ -566,8 +567,8 @@ class ILibraryAppletAccessor final : public ServiceFramework<ILibraryAppletAcces
 
 private:
     void GetAppletStateChangedEvent(Kernel::HLERequestContext& ctx) {
+        applet->GetBroker().SignalStateChanged();
```
Comment (Subv) [inline src/core/hle/service/am/am.cpp]: This changes behavior, why?

File: src/core/hle/kernel/hle_ipc.cpp [hunk 2]
```diff
@@ -36,11 +37,9 @@ void SessionRequestHandler::ClientDisconnected(const SharedPtr<ServerSession>& s
     boost::range::remove_erase(connected_sessions, server_session);
 }
 
-SharedPtr<Event> HLERequestContext::SleepClientThread(SharedPtr<Thread> thread,
-                                                      const std::string& reason, u64 timeout,
-                                                      WakeupCallback&& callback,
-                                                      Kernel::SharedPtr<Kernel::Event> event) {
-
+SharedPtr<WritableEvent> HLERequestContext::SleepClientThread(
```
Comment (Subv) [inline src/core/hle/kernel/hle_ipc.cpp]: I'm not sure I like this API. This lets you supply unlinked Readable and Writable events to the waiting machinery.

IMO it'd be better if you just took a WritableEvent and derived the ReadableEvent from it

File: src/core/hle/service/am/am.cpp (hunk 1 above)
Comment (DarkLordZach) [reply src/core/hle/service/am/am.cpp]: This did not change behavior, it just moved the signal to before the getter of the event, as the broker returns a readable event but exposes functions to signal the event. (original loc: https://github.com/yuzu-emu/yuzu/pull/1803/files#diff-c09cae2b75df143eebac772ac5616cf0L570)

File: src/core/hle/kernel/kernel.cpp [hunk 3]
```diff
@@ -17,9 +17,11 @@
 #include "core/hle/kernel/handle_table.h"
 #include "core/hle/kernel/kernel.h"
 #include "core/hle/kernel/process.h"
+#include "core/hle/kernel/readable_event.h"
```
Comment (lioncash) [inline src/core/hle/kernel/kernel.cpp]: These includes don't seem to be used.

**Comment by bunnei** [comment]
@DarkLordZach the build failures look real, FYI

**Comment by bunnei** [comment]
Needs a rebase

**Pull request merged**